src/sagemaker/image.py

```diff
@@ -0,0 +1,509 @@
+            instance_count (int): The number of instances to create.
+            image (str): docker image to use.
+            sagemaker_session (sagemaker.session.Session): a sagemaker session to use when interacting
+                with SageMaker.
+        """
+        from sagemaker.local_session import LocalSession
+        self.sagemaker_session = sagemaker_session or LocalSession()
+        self.instance_type = instance_type
+        self.instance_count = instance_count
+        self.image = image
+        self.hosts = ['{}-{}'.format(CONTAINER_PREFIX, i) for i in range(1, self.instance_count + 1)]
+        self.container_root = None
+        self.container = None
+        # set the local config. This is optional and will use reasonable defaults
+        # if not present.
+        self.local_config = None
+        if self.sagemaker_session.config and 'local'in self.sagemaker_session.config:
+            self.local_config = self.sagemaker_session.config['local']
+
+    def train(self, input_data_config, hyperparameters):
+        """Run a training job locally using docker-compose.
+        Args:
+            input_data_config (dict): The Input Data Configuration, this contains data such as the
+                channels to be used for training.
+            hyperparameters (dict): The HyperParameters for the training job.
+
+        Returns (str): Location of the trained model.
+
+        """
+        self.container_root = self._create_tmp_folder()
+
+        data_dir = tempfile.mkdtemp()
+        bucket_name = self.sagemaker_session.default_bucket()
+        volumes = []
+
+        # Set up the channels for the containers. For local data we will
+        # mount the local directory to the container. For S3 Data we will download the S3 data
+        # first.
+        for channel in input_data_config:
+            uri = channel['DataSource']['S3DataSource']['S3Uri']
+            parsed_uri = urlparse(uri)
+            key = parsed_uri.path.lstrip('/')
+
+            channel_name = channel['ChannelName']
+            channel_dir = os.path.join(data_dir, channel_name)
+            os.mkdir(channel_dir)
+
+            if uri.lower().startswith("s3://"):
+                self._download_folder(bucket_name, key, channel_dir)
+            else:
+                volumes.append(Volume(uri, channel=channel_name))
+
+        # Create the configuration files for each container that we will create
+        # Each container will map the additional local volumes (if any).
+        for host in self.hosts:
+            _create_config_file_directories(self.container_root, host)
+            self.write_config_files(host, hyperparameters, input_data_config)
+            shutil.copytree(data_dir, os.path.join(self.container_root, host, 'input', 'data'))
+
+        compose_data = self._generate_compose_file('train', additional_volumes=volumes)
+        compose_command = self._compose()
+        _stream_output(compose_command)
+        _cleanup()
```

Comment:
This cleanup and the other cleanup commands should be merged into a single cleanup method. This should be called in a finally block of this method.